File: dart/lib/src/token.dart
enum TokenType {
  illegal,
  eof,
  ident,
  int,
  // TODO: (Fa C Shus) consider changing equal to assign,
  equal,
  plus,
  comma,
  semicolon,
  lParen,
  rParen,
  lSquirly,
  rSquirly,
  function,
  let,
  bang,
  asterisk,
  slash,
  dash,
  gt,
  lt,
  eq,
  neq,
  true_,
  false_,
  if_,
  else_,
  return_;

  const TokenType();
}

class Token {
  final TokenType type;
  final String literal;

  const Token(this.type, this.literal);

  const Token.eof()
      : type = TokenType.eof,
        literal = "";

  const Token.illegal(this.literal) : type = TokenType.illegal;

  const Token.ident(this.literal) : type = TokenType.ident;

  const Token.int(this.literal) : type = TokenType.int;

  const Token.equal()
      : type = TokenType.equal,
        literal = "=";

  const Token.plus()
      : type = TokenType.plus,
        literal = "+";

  const Token.comma()
      : type = TokenType.comma,
        literal = ",";

  const Token.semicolon()
      : type = TokenType.semicolon,
        literal = ";";

  const Token.lParen()
      : type = TokenType.lParen,
        literal = "(";

  const Token.rParen()
      : type = TokenType.rParen,
        literal = ")";

  const Token.lSquirly()
      : type = TokenType.lSquirly,
        literal = "{";

  const Token.rSquirly()
      : type = TokenType.rSquirly,
        literal = "}";

  const Token.fn()
      : type = TokenType.function,
        literal = "fn";

  const Token.let()
      : type = TokenType.let,
        literal = "let";

  const Token.bang()
      : type = TokenType.bang,
        literal = "!";

  const Token.slash()
      : type = TokenType.slash,
        literal = "";

  const Token.dash()
      : type = TokenType.dash,
        literal = "-";

  const Token.asterisk()
      : type = TokenType.asterisk,
        literal = "*";

  const Token.eq()
      : type = TokenType.eq,
        literal = "==";

  const Token.neq()
      : type = TokenType.neq,
        literal = "!=";

  const Token.gt()
      : type = TokenType.gt,
        literal = ">";

  const Token.lt()
      : type = TokenType.lt,
        literal = "<";

  const Token.true_()
      : type = TokenType.true_,
        literal = "true"; 

  const Token.false_()
      : type = TokenType.false_,
        literal = "false";  

  const Token.if_()
      : type = TokenType.if_,
        literal = "if";

  const Token.else_()
      : type = TokenType.else_,
        literal = "else";

  const Token.return_()
      : type = TokenType.return_,
        literal = "return";
        

  @override
  String toString() => "Token.${type.name}('$literal')";

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Token && type == other.type && literal == other.literal;

  @override
  int get hashCode => type.hashCode ^ literal.hashCode;
}

File: cpp-spongman/src/lexer/token.cpp
#include <ostream>
#include <sstream>

#include "token.hpp"


std::ostream& operator<<(std::ostream& os, const TokenType& tokenType)
{
	switch (tokenType) {
		case TokenType::Illegal:        os << "(illegal)"; break;
		case TokenType::Eof:            os << "(eof)"; break;
		case TokenType::Identifier:     os << "(identifier)"; break;

		case TokenType::Integer:        os << "(integer)"; break;
		case TokenType::Comma:          os << ","; break;
		case TokenType::Semicolon:      os << ";"; break;
		case TokenType::Dollar:         os << "$"; break;
		case TokenType::Colon:          os << ":"; break;
		
		case TokenType::Lparen:         os << "("; break;
		case TokenType::Rparen:         os << ")"; break;
		case TokenType::Lsquirly:       os << "{"; break;
		case TokenType::Rsquirly:       os << "}"; break;
		case TokenType::Lbracket:       os << "["; break;
		case TokenType::Rbracket:       os << "]"; break;

		// OpTokenType::erator:         os << ""; break;
		case TokenType::Assign:         os << "="; break;
		case TokenType::Plus:           os << "+"; break;
		case TokenType::Minus:          os << "-"; break;
		case TokenType::Bang:           os << "!"; break;
		case TokenType::Asterisk:       os << "*"; break;
		case TokenType::Slash:          os << "/"; break;
		case TokenType::Percent:        os << "%"; break;
		case TokenType::Tilde:          os << "~"; break;
		case TokenType::BitAnd:         os << "&"; break;
		case TokenType::BitOr:          os << "|"; break;
		case TokenType::BitEor:         os << "^"; break;
		case TokenType::And:            os << "&&"; break;
		case TokenType::Or:             os << "||"; break;

		case TokenType::Lt:             os << "<"; break;
		case TokenType::Gt:             os << ">"; break;
		case TokenType::Le:             os << "<="; break;
		case TokenType::Ge:             os << ">="; break;

		// KeTokenType::yword:          os << ""; break;
		case TokenType::Function:       os << "fn"; break;
		case TokenType::Let:            os << "let"; break;
		case TokenType::True:           os << "true"; break;
		case TokenType::False:          os << "false"; break;
		case TokenType::If:             os << "if"; break;
		case TokenType::Else:           os << "else"; break;
		case TokenType::Return:         os << "return"; break;

		case TokenType::Eq:             os << "=="; break;
		case TokenType::Not_eq:         os << "!="; break;

		case TokenType::String:         os << "(String)"; break;
		default:
			os << "(unhandled " << static_cast<int>(tokenType) << ")";	
			break;
	}
	return os;
}

std::ostream& operator<<(std::ostream& os, const Token& token)
{
	switch (token.type) {
		case TokenType::Identifier: os << token.literal; break;
		case TokenType::Integer:    os << token.literal; break;
		case TokenType::String:     os << "\"" << token.literal << "\""; break;
		default:                    os << token.type; break;
	}
	return os;
}


std::string std::to_string(TokenType tokenType)
{
	std::stringstream ss;
	ss << tokenType;
	return ss.str();
}

std::string std::to_string(const Token& token)
{
	std::stringstream ss;
	ss << token;
	return ss.str();
}
